File: app/src/main/java/it/unical/mat/embasp/languages/pddl/Plan.java
package it.unical.mat.embasp.languages.pddl;

import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import it.unical.mat.embasp.base.Output;

/** A simplified solution to a PDDL problem */

public abstract class Plan extends Output {
	protected List<Action> actionSequence;

	private List<Object> actionsObjects;

	public Plan(final String plan, final String error) {
		super(plan, error);
	}

	public List<Action> getActions() {
		if (actionSequence == null) {
			actionSequence = new ArrayList<>();
			parse();
		}

		return Collections.unmodifiableList(actionSequence);
	}

	public List<Object> getActionsObjects() throws InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException {
		if (actionsObjects == null) {
			actionsObjects = new ArrayList<>();
			final PDDLMapper mapper = PDDLMapper.getInstance();
			for (final Action a : getActions()) {
				final Object obj = mapper.getObject(a.getName());
				if (obj != null)
					actionsObjects.add(obj);
			}
		}

		return actionsObjects;
	}
}

File: CSharp version/it/unical/mat/embasp/languages/pddl/Plan.cs
using it.unical.mat.parsers.pddl;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace it.unical.mat.embasp.languages.pddl
{
    using Output = it.unical.mat.embasp.@base.Output;

    public abstract class Plan : Output, IPDDLDataCollection
    {
        protected internal IList<Action> actionSequence;
        private IList<object> actionsObjects;

        public Plan(string plan, string error) : base(plan, error) { }

        public virtual IList<Action> Actions
        {
            get
            {
                if (actionSequence == null)
                {
                    actionSequence = new List<Action>();
                    Parse();
                }
                return new ReadOnlyCollection<Action>(actionSequence);
            }
        }

        public virtual IList<object> ActionsObjects
        {
            get
            {
                if (actionsObjects == null)
                {
                    actionsObjects = new List<object>();
                    PDDLMapper mapper = PDDLMapper.Instance;
                    //Console.WriteLine("OK");
                    foreach (Action a in Actions)
                    {
                        object obj = mapper.GetObject(a.Name);
                        //Console.WriteLine(obj.ToString());
                        if (obj != null)
                            actionsObjects.Add(obj);
                    }
                }
                //Console.WriteLine(actionsObjects.ToString());
                return actionsObjects;
            }
        }

        public void StoreAction(string action)
        {
            actionSequence.Add(new Action(action));
        }
    }
}